### ISSUE TITLE
Import OpenTofu as a library to Go application

### ISSUE THREAD

**Issue by nabokihms** [opentofu/opentofu#810]
### OpenTofu Version

```shell
Any version.
```

### Use Cases

This is the feature awaited by many developers in the terraform community for a very long time.

My case is simple. We have a tool (CLI) that provisions cloud infrastructure using terraform under the hood. After provisioning the infrastructure, it installs Kubernetes on top of it and provisions other software. 

The tool has pluggable modules that can be added in runtime (that is why Pulumy is not applicable 😅 ). The only officially supported way is to execute terraform as a binary. However, to make it work, users should install terraform (of the right version) first, and then install our tool, which ruins the user experience completely.

Many applications already utilize terraform the same way under the hood to provision cloud infrastructure, e.g., [openshift/installer](https://github.com/openshift/installer).

### Attempted Solutions

There is a terraform-exec library: https://github.com/hashicorp/terraform-exec (which is still under MPL). 
Instead of executing commands from the code, it utilizes the terraform binary, which leads to many disadvantages.

1. You need to install Terraform.
2. There is no analog for tofu, and it requires you to execute tofu using the terraform library.
3. Using `exec` is slower, harder to parallel, manage state, etc.

### Proposal

Make a public interface package that will be kept and maintained as a library.
This package should include methods to invoke all tofu commands like `init`, `plan`, `apply`, `destroy`, `output`, and others.

P.S. I hope that OpenTofu, which is a community-driven project, is open to such changes.

### References

https://github.com/hashicorp/terraform/issues/12582

**Comment by Adesoji1** [comment]
Hi @nabokihms,

Thank you for taking the time to submit this feature request. Your proposal to create a public interface for OpenTofu to be used as a library in Go applications is an interesting idea that could potentially streamline workflows for many developers.

We understand the importance of a smooth user experience and the challenges of managing dependencies like Terraform versions. The ability to directly invoke operations like `init`, `plan`, `apply`, `destroy`, and `output` through a Go library rather than through the command line or by managing a separate Terraform binary could indeed be beneficial.

Before we proceed, there are a few considerations we need to address:

1. **Scope and Maintenance**: Developing and maintaining a library interface would require careful planning to ensure that it aligns with OpenTofu's roadmap and that we have the resources to maintain it long-term.

2. **Compatibility**: We need to evaluate how such an interface would remain compatible with different versions of Terraform and whether it could adapt to the Terraform plugin protocol's changes.

3. **Community Feedback**: It is essential to gather more feedback from the community. This would help us determine how many users would benefit from this feature and prioritize it accordingly.

4. **Implementation Details**: We would need to discuss further the architecture of this library. For instance, whether it should be a thin wrapper around the Terraform CLI or a more robust integration with OpenTofu's core functionalities.

Because for me this is what i am thinking in my head though :smiley_cat:  :roll_eyes: 

package opentofu :smile: 

import (
    "os/exec"
    "strings"
)

// TerraformCmd executes a Terraform command and returns its output or an error.
func TerraformCmd(args ...string) (string, error) {
    cmd := exec.Command("terraform", args...)
    output, err := cmd.CombinedOutput()
    return string(output), err
}

// Init initializes a new or existing Terraform configuration.
func Init(options ...string) (string, error) {
    return TerraformCmd(append([]string{"init"}, options...)...)
}

// Apply applies the Terraform plan.
func Apply(options ...string) (string, error) {
    return TerraformCmd(append([]string{"apply"}, options...)...)
}

// Destroy destroys the Terraform-managed infrastructure.
func Destroy(options ...string) (string, error) {
    return TerraformCmd(append([]string{"destroy"}, options...)...)
}

// Other functions for plan, output, etc. would follow a similar pattern...


6. **Security**: We need to consider how the library would handle state management, authentication, and other sensitive data without compromising security.

These are the thoughts in my :brain:  , does this make sense?

**Comment by Magnitus-** [comment]
I think this idea is similar to this one: https://github.com/opentofu/opentofu/issues/758

While one suggest a C library and this one a go package (and also personally more inclined to want a go package), I think the ideas are somewhat similar.

Note that opentofu has forked terraform-exec already, though I'm not sure to what are the long term plans of the community with this (interest seems low, though ours is pretty high as we are using terraform-exec for our cd solution built on top of terraform): https://github.com/opentofu/tofu-exec

I've used terraform-exec for our terraform pipeline tool and found it convenient to use (its not a big overhead to stick an extra terraform binary in a docker image or on an automation server) although yes, its a bit less nice if you want to sell something to third parties that is packaged as a single self-contained binary (though so far, we've found the ability to decouple the versioning of terraform from the versioning of our cd tool to be a plus).

In terms of performance, calling terraform init + terraform plan + terraform apply, plus having all the communication with providers over grpc as well as the backend (for the state) and waiting on various apis we are interacting with to respond is already an overhead of its own so we haven't really noticed a lot of additional overhead on top of that from calling a separate process. That's on an automation servers managing about 30-50 pipelines getting called every minute or so.

The one thing I could see potentially biting use as we get more fancy may be the processing of the logs (haven't really delved into log analytics much yet), but we'll see.

For now, for us, just having a direct port of terraform-exec would already be a bit step (having some kind of programmer-friendly api to manage opentofu from) that we could go forward with.

Perhaps it would make sense to start with that, make sure we have an api contract that is mature with the separate binary solution and from there, maybe the same api contract could be used for a direct library, a grpc api and/or something else.

For developers, it would be fairly easy to move from one backend implementation to another as long as the api contract doesn't change too much.

**Comment by dvaumoron** [comment]
Hi @Adesoji1, 

I dont get your 4, using a wrapper calling an other program doesn't seem to be the goal here, a direct access with embedding to avoid a separate installation seem more accurate, but that's more work to adapt the internal code to allow an easy usage as library.

**Comment by dvaumoron** [comment]
After a quick check ont the code, writing a wrapper to internal part need at most to start by changing the meta.View type to an interface (to allow to retrieve data and return them instead of printing them), but the refactor to use View is not done everywhere (as stated in #420 ), some part still rely on meta.Ui wich doesn't even use meta.Streams and rely on the os.Std*

So depending on which part is needed as library, the refactor would have to be done first in it (the init command is one of those)

**Comment by dvaumoron** [comment]
To be clearer, changing the meta.View will allow the library to return a tfdiags.Diagnostics to the caller (this type and others need to be moved outside of the internal part)

**Comment by cube2222** [comment]
> We understand the importance of a smooth user experience and the challenges of managing dependencies like Terraform versions.

Just a quick note to avoid confusion and clarify, @Adesoji1 is not part of the core engineering team, so I’m assuming the “we” here means “we as the community”.

---

Regarding the proposal, I think it makes sense, but there are some problems to consider. The ones I see off the top of my head:
- Versioning: Tools that underneath run other tools very often want to support many versions of those underlying tools. With a library you're generally bound to the version you statically linked with.
- Providers: Providers right now are separate binaries that are downloaded and managed on the local filesystem, executed as separate processes.
- API stability: We'd probably want a library layer where we can promise a moderate level of stability, that vaguely corresponds to the current CLI commands and arguments.

**Comment by cube2222** [comment]
Additional thought - if you still need to fetch the providers when running init, then why not just download tofu itself as well?

**Comment by lplazas** [comment]
This is an extra layer deep, but providers could be distributed as go packages themselves, so initializing your go project with opentofu+providers would be taken care of by go modules.

**Comment by cube2222** [comment]
This seems to be a duplicate of #758

**Comment by nabokihms** [comment]
@cube2222 unfortunate to hear. 
- By saying we, you mean we as maintainers or we as a whole community?
- By saying foreseeable future you mean that implementation of this feature can still be considered in the future?

I’d like to keep the issue open then to have a place for people to land and share their usecases. I still believe that with the support of the community even the involved complexity to deal with will be a piece of cake.

**Comment by dvaumoron** [comment]
After a look at tofu-exec, it seems that it already allows to install a tofu version before launching it...

**Comment by janosdebugs** [comment]
@nabokihms we as in the core maintainers. We would love if we were in a position where this was only a matter of deciding when to do it based on the use cases, but it unfortunately isn't.

From my experience working on OpenTofu, the OpenTofu codebase is huge and has a lot of globally scoped code, which means it doesn't lend itself for being used as a library. You wouldn't be able to call `plan` or `apply` twice in a row with different parameters as they would have adverse effects on each other. Restarting the application between runs is, unfortunately, a hard requirement with the current code. Until the codebase (currently roughly 300.000 lines of code) have been thoroughly refactored and global scope is removed, it is not possible to provide a library. Considering the amount of code involved and the backwards compatibility needs involved this is a very long undertaking and may never materialize.

**Comment by minamijoyo** [comment]
@cube2222 @janosdebugs One of the goals of the OpenTofu project is "Layered and modular". I am wondering what this actually means.

https://opentofu.org/

<img width="1368" alt="image" src="https://github.com/opentofu/opentofu/assets/6985802/66a196ae-6a89-4b62-92fc-f78b49d54bd6">

> Layered and modular
with a programmer-friendly project structure to encourage building on top, enabling a new vibrant ecosystem of tools and integrations.

I had expected this meant exposing OpenTofu's internals as a collection of reusable Go libraries. Can you clarify what exactly this means?

**Comment by janosdebugs** [comment]
Hey @minamijoyo it's a goal to get there, but as much as it would be awesome to run OpenTofu as a library, it is not possible for everything today.

There are several parts of OpenTofu that are realistically modularizable today. While not decided yet, for example there's #382, which would allow people to write their own state backends, or #1138 which would (among others) make it realistic to write and distribute providers in other, non-statically compiled programming languages. As we make large parts of the codebase work as plugins, we'll also get rid of globally scoped code. At the end of that road, hopefully, lies being able to expose OpenTofu as a nice, reusable Go API that you can use to embed OpenTofu and use it the same way you would from the CLI.

So why are we not leaving the issue open as a goal?

1. There's a lot of code to work through and a lot that needs to happen before setting this as a goal is realistic.
2. Unless we are grossly mistaken, there's nothing that can be done to make this process any faster other than work on modularizing OpenTofu piece by piece. This is not a question of not enough community upvotes, or need, it's very clear that this is needed and wanted. This will not happen in the near future, so waiting for it is unfortunately not fruitful.
3. When OpenTofu is modularized all the way through, this will happen automatically. However, as said before, this will take a very long time.
4. There is a workaround in the form of [tofu-exec](https://github.com/opentofu/tofu-exec) today.

**Comment by minamijoyo** [comment]
@janosdebugs I appreciate your detailed explanation and respect the core team's decision. Thanks!

**Comment by siennathesane** [comment]
> There's a lot of code to work through and a lot that needs to happen before setting this as a goal is realistic.

I think setting this as an explicit goal is actually the right thing to do - especially when your opening statement is _"it's a goal to get there"_. With it as a measurable goal, it surfaces the amount of effort required to achieve this desired outcome. It also gives folks a way to ask questions about the work that needs to be done, and surface opportunities to make this happen more effectively _with more potential contributions from the community_, which was the whole purpose of this project.

> This is not a question of not enough community upvotes, or need, it's very clear that this is needed and wanted. This will not happen in the near future, so waiting for it is unfortunately not fruitful.

Many of us actively waited on this from Hashicorp for years, and I personally built a private fork of Terraform to work around this limitation. If this is going to be community-driven software that's actually modular, we the community need to know the work required achieve the goals that we the community want from this project.

Will it happen quickly? No, and that's okay. But it definitely needs to be a goal that can be measured against because while _waiting for it is [...] not fruitful_, **we are actively waiting on it.**

Does that mean this issue needs to be reopened? I don't know, I'm not part of the project team. But this definitely needs to be something that we the community can come to as a point of reference to see where it's at.
